addons/crm/models/crm_lead.py

```diff
@@ -393,11 +410,37 @@ def _compute_contact_name(self):
         for lead in self:
             lead.update(lead._prepare_contact_name_from_partner(lead.partner_id))
 
+    @api.depends('partner_id')
+    def _compute_partners(self):
+        self.company_partner_id = False
+        self.contact_partner_id = False
+        for lead in self.filtered('partner_id'):
+            if lead.partner_id.is_company:
+                lead.company_partner_id = lead.partner_id
+            else:
+                lead.contact_partner_id = lead.partner_id
+                lead.company_partner_id = lead.partner_id.parent_id
+
+    def _inverse_partners(self):
```

Comment:
Looks like this solution is much better as fields are not stored. It is also more readable. Maybe we could also handle the case we first create the contact and then set a company, because atm it will be ignored. Probably to check with LNA as well, no need to make the logic too complex